feat: allow manual repair edit

leonardmq opened the pull request:
## What does this PR do?

This PR adds support for manually editing a task run repair. After `Attempt repair`, the user has a new button (`Edit`) that turns the repair output display into a text area that allows users editing the repair attempt.

This PR also fixes the missing validation of the repair output in the repair API. There was a bug that allowed repair outputs that do not match the structured output schema, which might have caused downstream problems due to corrupt shapes.

This is an acceptable solution but it has a major limitation due to requiring the user to first `Attempt (automatic) repair`. In the case of a `TaskRun` accessed from the `Dataset` page, the user does not have the ability to select the model used during the repair - it is by default the same as the one that generated the original run output. If a run used a small model (e.g. the small Llama models), the user may be stuck as the small model may fail to even pass validation during `Attempt repair`, which means the user then never gets access to the `Edit`.

New `Edit` button:
<img width="746" alt="image" src="https://github.com/user-attachments/assets/cdb65979-b46d-4079-bde0-3c736cb53bee" />

`Edit` mode:
<img width="752" alt="image" src="https://github.com/user-attachments/assets/d4e69274-3e30-4949-992b-774c833c6251" />

Error state (if JSON does not match schema):
<img width="757" alt="image" src="https://github.com/user-attachments/assets/120cfe7c-5e7b-453f-b025-f0fab478b9ad" />


## Related Issues

Manual edit requested: https://github.com/Kiln-AI/Kiln/issues/115

## Contributor License Agreement

I, @leonardmq, confirm that I have read and agree to the [Contributors License Agreement](https://github.com/Kiln-AI/Kiln/blob/main/CLA.md).

## Checklists

- [x] Tests have been run locally and passed
- [x] New tests have been added to any work in /lib

leonardmq commented:
WIP because I want to add a couple of tests in the repair API

leonardmq commented on app/desktop/studio_server/repair_api.py:
```diff
@@ -73,6 +76,12 @@ async def post_repair_run(
         task, run = task_and_run_from_id(project_id, task_id, run_id)
+
+        # the repair must match the task's output schema if structured output is enabled
+        if task.output_json_schema:
```
I have not found a centralized utility that could be used to do the validation for this. The validation in the adapter does not seem to be usable outside the context of a generation run

leonardmq commented on app/web_ui/src/routes/(app)/run/run.svelte:
```diff
@@ -343,6 +343,25 @@
+    if (!repair_run) {
+      return
+    }
+    repair_run.output.output = repair_output_edited
```
The way it currently works is that the repair run would still be attributed to whichever model generated the repair output that was repaired. Would you rather want the repair run to be different?

scosman replied:
yeah, we should attribute it to the person making the edit. A standard Human datasource should work!

scosman commented:
ohhh, can't wait!

Let's pretty print for the text-area if it's JSON? Makes it easier to edit. Basically: `JSON.stringify(obj, null, 2)`

leonardmq commented on app/web_ui/src/routes/(app)/run/output_repair_edit_form.svelte:
```diff
@@ -0,0 +1,116 @@
+  function validate_output(text: string) {
+    if (task.output_json_schema) {
+      try {
+        // TODO: validate against the JSON schema instead of just checking if it is valid JSON
```
@scosman - I'd want to validate the schema on the client-side here, but the webapp does not have a library for JSON schema validation. 

Is there any particular library you would prefer to use for JSON schema validation on the client side?

scosman replied:
We can leave validation to server! We just shouldn't save it without validation, and should make a decent error message there.

leonardmq replied:
I added validation on the server side, but doing it on the client too would be nice as it allows validating in realtime and doing things like disabling the submit button while invalid. 

The current UI for editing the repair requires that you first save to apply the edit to the unsaved repair, and then you submit the repair. If we do server-only, the validation feedback would not be on this same form. 

I could change it to make the edit form directly submit to the API (instead of applying to the run without submitting) if you think that makes more sense - then server validation errors could come through on this same form

leonardmq commented on app/web_ui/src/routes/(app)/run/output_repair_edit_form.svelte:
```diff
@@ -0,0 +1,116 @@
+      output: {
+        ...repair_run.output,
+        output: repair_output_edited,
+        source: {
```
I tried `source: null` and `source: undefined` instead of prefilling like this, but the API validation does not allow that

leonardmq commented on app/web_ui/src/routes/(app)/run/output_repair_edit_form.svelte:
```diff
@@ -0,0 +1,116 @@
+      <div
+        class="text-xs font-medium text-error flex flex-row mb-2 items-center"
+      >
+        <svg
```
This SVG comes from the `Warning` component

leonardmq commented on app/web_ui/src/routes/(app)/run/output_repair_edit_form.svelte:
```diff
@@ -0,0 +1,116 @@
+  {#if task.output_json_schema}
+    {#if repair_output_edited_valid}
+      <div class="text-xs font-medium text-gray-500 flex flex-row mb-2">
+        <svg
```
This SVG is the same used elsewhere in the page to show `Structure Valid`.

Maybe worth wrapping each SVG into components and giving them names

scosman replied:
totally. I try to do that when I duplicate it.... but as you can tell I'm not religious about it

leonardmq commented:
@scosman 

Added the following changes:
- add a `disabled` state on the `FormContainer` to prevent submitting
- prevent submitting form if the `submit` button is not visible
- add server-side validation of the repair output if the task has structured output
- refactor repair edit form:
  - use the `FormContainer`
  - prettify initial value
  - add client-side validation if the task has structured output
  - sets the `human` source (prefilled on frontend, and user id is set on the server side)
- test for manual repair source

---

https://github.com/user-attachments/assets/e24f1673-631a-4f33-8eb6-0bcea4fac52f

scosman commented:
This looks great. Few quick comments, but I'll be able to look in more depth next week!

scosman commented:
@leonardmq is this ready for me to take a look?

leonardmq commented:
> @leonardmq is this ready for me to take a look?

Yes should be ready - some open questions around desired behavior of the UI, so if changes are wanted there, I'll tweak it after the review

scosman commented on app/web_ui/src/routes/(app)/run/run.svelte:
```diff
@@ -425,15 +445,24 @@
               />
             </FormContainer>
-          {:else if repair_review_available}
+          {:else if repair_review_available || repair_edit_mode}
```
Let's just make these two separate blocks? They are quite different. While we're using the repair to seed the manual, the user may completely change it. Text like "The model has attempted to fix the output given" becomes out of date.

We can add a paragraph with explanation text in new block, something like "Manually improve or correct the response."

leonardmq replied:
✅ - split those two into separate blocks and used `Manually improve or correct the response.` as text at the top

scosman commented on app/web_ui/src/routes/(app)/run/run.svelte:
```diff
@@ -459,24 +488,29 @@
+
             {#if accept_repair_error}
-              <p class="text-error font-medium text-sm">
+              <p class="mt-2 text-error font-medium text-sm">
```
this is my fault, but error should always be above the buttons. If not, they can appear off screen depending on where the user scrolls to, and make the UI feel like nothing happened.

leonardmq replied:
I have not moved this error block because not sure exactly where you would want them to be. But I have moved the errors above the buttons in my manual repair form component

scosman reviewed (changes requested):
Looks good, some small comments. Like your UX suggestion for the save flow.

Issue where if auto-repair fails, you can’t manual repair: I’m okay with this for now. If the selected model fails repeatedly, you have bigger problems. This problem is not being introduced here, repairing would already be failing: Let’s see if this is a common complaint before fixing.

leonardmq commented on app/web_ui/src/routes/(app)/run/output_repair_edit_form.svelte:
```diff
@@ -0,0 +1,116 @@
+  {/if}
+
+  <FormContainer
+    submit_label="Apply"
```
✅  - renamed

leonardmq commented on app/web_ui/src/routes/(app)/run/output_repair_edit_form.svelte:
```diff
@@ -0,0 +1,116 @@
+
+    <button
+      class="link text-gray-500 text-sm text-right"
+      on:click={handle_cancel}>Cancel edit</button
```
✅  - renamed

leonardmq commented on app/web_ui/src/routes/(app)/run/output_repair_edit_form.svelte:
```diff
@@ -0,0 +1,116 @@
+  >
+    <FormElement
+      id={"repair_manual_output"}
+      label="Manual Repair"
```
✅ - edited the text

leonardmq commented:
@scosman 

Pushed changes to align with your review comments:
- Refactored to submit directly from the repair edit form (removing intermediate state)
- Removed all client-side validation—including the `Malformed JSON` warning and the disabled state on the submit button—to keep validation fully server-side for simplicity and consistency
- Updated `validate_repaired_output` in `TaskRun` to call `validate_schema` on the provided repair output (if any), when the task has an output JSON schema; added a couple of tests for this
- Removed TODOs and tweaked some of the UI text

The error message from the validator might still be a bit off, since it comes from `validate_schema`, which currently has a narrow message format. I opened a separate PR to address that: https://github.com/Kiln-AI/Kiln/pull/272 - since fixing it required touching unrelated parts of the codebase.

scosman commented:
👍👍👍

scosman reviewed (approved):
👍
